homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -0,0 +1,400 @@
+        return cast(dict, new_token)
+
+    async def async_refresh_token(self, token: dict) -> dict:
+        """Refresh tokens."""
+        session = async_get_clientsession(self.hass)
+
+        data = {
+            "grant_type": "refresh_token",
+            "client_id": self.client_id,
+            "refresh_token": token["refresh_token"],
+        }
+
+        if self.client_secret is not None:
+            data["client_secret"] = self.client_secret
+
+        resp = await session.post(self.token_url, json=data)
+
+        resp.raise_for_status()
+
+        new_token = await resp.json()
+        new_token["expires_at"] = time.time() + new_token["expires_in"]
+
+        return {**token, **new_token}
+
+
+class AbstractOAuth2FlowHandler(config_entries.ConfigFlow, metaclass=ABCMeta):
+    """Handle a config flow."""
+
+    DOMAIN = ""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_UNKNOWN
+
+    def __init__(self) -> None:
+        """Instantiate config flow."""
+        if self.DOMAIN == "":
+            raise TypeError(
+                f"Can't instantiate class {self.__class__.__name__} without DOMAIN being set"
+            )
+
+        self.external_data: Any = None
+        self.flow_impl: AbstractOAuth2Implementation = None  # type: ignore
+
+    @property
+    @abstractmethod
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+
+    async def async_step_pick_implementation(self, user_input: dict = None) -> dict:
+        """Handle a flow start."""
+        assert self.hass
+        implementations = async_get_implementations(self.hass, self.DOMAIN)
+
+        if not implementations:
+            return self.async_abort(reason="missing_configuration")
+
+        # Pick first implementation as we have only one.
+        self.flow_impl = list(implementations.values())[0]
+        return await self.async_step_auth()
+
+    async def async_step_auth(self, user_input: dict = None) -> dict:
+        """Create an entry for auth."""
+        # Flow has been triggered by external data
+        if user_input:
+            self.external_data = user_input
+            return self.async_external_step_done(next_step_id="creation")
+
+        try:
+            with async_timeout.timeout(10):
+                url = await self.flow_impl.async_generate_authorize_url(self.flow_id)
+        except asyncio.TimeoutError:
+            return self.async_abort(reason="authorize_url_timeout")
+
+        return self.async_external_step(step_id="auth", url=url)
+
+    async def async_step_creation(self, user_input: dict = None) -> dict:
+        """Create config entry from external data."""
+        tokens = await self.flow_impl.async_resolve_external_data(self.external_data)
+
+        self.logger.info("Successfully authenticated")
+
+        return await self.async_oauth_create_entry(
+            {"implementation": self.flow_impl.domain, "token": tokens}
```

Comment:
Can we rename `"implementation"` to `"integration"`?

Why do we need to store the domain for the implementation? Don't we know the integration domain already?

Can we replace the domain with some other id to be able to have more than one implementation per domain?

Reply:
I went for the word `implementation` because that's how I was calling it inside the OAuth2 helper. `integration` is too generic, and then you don't know what it is about, especially confusing because it is inside a config entry that has their own domain.

It will be possible that an implementation won't be provided by the integration itself. I am working on an account linking service that will allow Home Assistant to have its own client ID/secret that users can use. That implementation will be provided by another integration.

I don't see the use case for more than one integration per domain. It would also bring the trouble with it that now each implementation will need to generate its own identifier. I think a lot of extra work for no benefit.

Reply:
Ok. "implementation" is also pretty generic although I see the origin. Is "oauth_implementation" too long?

Reply:
I'll change it to `auth_implementation`.